src/envoy/http/metadata_exchange/config.cc

```diff
@@ -37,51 +37,79 @@ using Common::Wasm::Null::Plugin::replaceRequestHeader;
 
   mcs.SetSerializationDeterministic(true);
   if (!metadata.SerializeToCodedStream(&mcs)) {
-    logWarn(
-        absl::StrCat("unable to serialize Nodemetadata key=", NodeMetadataKey));
+    logWarn("unable to serialize metadata");
     return false;
   }
   return true;
 }
 
-void PluginRootContext::onConfigure(
-    std::unique_ptr<WasmData> ABSL_ATTRIBUTE_UNUSED configuration) {
-  google::protobuf::Value metadata;
-  if (getMetadataValue(Common::Wasm::MetadataType::Node, NodeMetadataKey,
-                       &metadata) != Common::Wasm::MetadataResult::Ok) {
-    logWarn(absl::StrCat("cannot get metadata for: ", NodeMetadataKey));
+}  // namespace
+
+void PluginRootContext::updateMetadataValue() {
+  google::protobuf::Value keys_value;
+  if (getMetadataValue(Common::Wasm::MetadataType::Node,
+                       NodeMetadataExchangeKeys,
+                       &keys_value) != Common::Wasm::MetadataResult::Ok) {
+    logWarn(
+        absl::StrCat("cannot get metadata key: ", NodeMetadataExchangeKeys));
     return;
   }
 
-  if (metadata.kind_case() == google::protobuf::Value::kStructValue) {
-    std::string metadata_bytes;
-    serializeToStringDeterministic(metadata.struct_value(), &metadata_bytes);
-
-    metadata_value_ =
-        Base64::encode(metadata_bytes.data(), metadata_bytes.size());
+  if (keys_value.kind_case() != google::protobuf::Value::kStringValue) {
+    logWarn(absl::StrCat("metadata key is not a string: ",
+                         NodeMetadataExchangeKeys));
+    return;
+  }
 
-    // magic "." to get the whole node.
-    google::protobuf::Struct node;
-    if (getMetadataStruct(Common::Wasm::MetadataType::Node, WholeNodeKey,
-                          &node) != Common::Wasm::MetadataResult::Ok) {
-      logWarn(absl::StrCat("cannot get metadata for: ", WholeNodeKey));
-      return;
+  google::protobuf::Struct metadata;
+
+  // select keys from the metadata using the keys
+  const std::set<absl::string_view> keys =
+      absl::StrSplit(keys_value.string_value(), ',', absl::SkipWhitespace());
+  for (auto key : keys) {
+    google::protobuf::Value value;
+    if (getMetadataValue(Common::Wasm::MetadataType::Node, key, &value) ==
+        Common::Wasm::MetadataResult::Ok) {
+      (*metadata.mutable_fields())[std::string(key)] = value;
+    } else {
+      logWarn(absl::StrCat("cannot get metadata key: ", key));
     }
+  }
+
+  // store serialized form
+  std::string metadata_bytes;
+  serializeToStringDeterministic(metadata, &metadata_bytes);
+  metadata_value_ =
+      Base64::encode(metadata_bytes.data(), metadata_bytes.size());
+}
+
+void PluginRootContext::onConfigure(
+    std::unique_ptr<WasmData> ABSL_ATTRIBUTE_UNUSED configuration) {
+  updateMetadataValue();
 
+  // TODO: this is really expensive since it fetches the entire metadata from
+  // before magic "." to get the whole node.
+  google::protobuf::Struct node;
+  if (getMetadataStruct(Common::Wasm::MetadataType::Node, WholeNodeKey,
```

Comment:
You could use namedMetadataValue which would cache the data and use a hash table to return the field you want.

Reply:
Caching is worse here since it cause extra memory cost, and there's only one call, at start-up.